hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/autocleanup/AutoCleanupScheduler.java

```diff
@@ -0,0 +1,98 @@
+
+/**
+ * A scheduler to invoke a set of cleanup handlers periodically.
+ */
+public class AutoCleanupScheduler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AutoCleanupScheduler.class);
+
+    private static final String AUTO_CLEANUP = "auto-cleanup";
+    private static final String SEP = ".";
+    private static final String PROP_AUTO_CLEANUP_INTERVAL = "${hawkbit.autocleanup.scheduler.fixedDelay:86400000}";
+
+    private final SystemManagement systemManagement;
+    private final SystemSecurityContext systemSecurityContext;
+    private final LockRegistry lockRegistry;
+    private final List<CleanupTask> cleanupTasks;
+
+    /**
+     * Constructs the cleanup schedulers and initializes it with a set of
+     * cleanup handlers.
+     * 
+     * @param systemManagement
+     *            Management APIs to invoke actions in a certain tenant context.
+     * @param systemSecurityContext
+     *            The system security context.
+     * @param lockRegistry
+     *            A registry for shared locks.
+     * @param cleanupTasks
+     *            A list of cleanup tasks.
+     */
+    public AutoCleanupScheduler(final SystemManagement systemManagement,
+            final SystemSecurityContext systemSecurityContext, final LockRegistry lockRegistry,
+            final List<CleanupTask> cleanupTasks) {
+        this.systemManagement = systemManagement;
+        this.systemSecurityContext = systemSecurityContext;
+        this.lockRegistry = lockRegistry;
+        this.cleanupTasks = cleanupTasks;
+    }
+
+    /**
+     * Scheduler method which retrieves all tenants from
+     * {@link SystemManagement#findTenants()} and runs all registered auto
+     * cleanup handlers for each of them.
+     */
+    @Scheduled(initialDelayString = PROP_AUTO_CLEANUP_INTERVAL, fixedDelayString = PROP_AUTO_CLEANUP_INTERVAL)
+    public void run() {
+        LOGGER.debug("Auto cleanup scheduler has been triggered.");
+        // run this code in system code privileged to have the necessary
+        // permission to query and create entities
+        if (!cleanupTasks.isEmpty()) {
+            systemSecurityContext.runAsSystem(this::executeAutoCleanup);
+        }
+    }
+
+    @SuppressWarnings("squid:S3516")
+    private Void executeAutoCleanup() {
+        systemManagement.forEachTenant(tenant -> cleanupTasks.forEach(task -> {
+            final Lock lock = obtainLock(task, tenant);
+            if (!lock.tryLock()) {
+                return;
+            }
+            try {
+                task.run();
+            } catch (final RuntimeException e) {
+                LOGGER.error("Cleanup task failed.", e);
+            } finally {
+                lock.unlock();
+            }
+        }));
+        return null;
+    }
+
+    private Lock obtainLock(final CleanupTask task, final String tenant) {
+        final String lockId = new StringBuilder().append(AUTO_CLEANUP).append(SEP).append(task.getId()).append(SEP)
```

Comment:
According to my IDE, using `String` instead of `StringBuilder` is more efficient in this case.